### ISSUE TITLE
Timeseries forecasting tutorial has bias issues

### ISSUE THREAD

**Issue by SarnathK** [tensorflow/tensorflow#40139]
## URL(s) with the issue:

https://www.tensorflow.org/tutorials/structured_data/time_series
### Clear description

In the time-series forecasting tutorial, the normalization is done prior to obtaining time-series windows. 
Consider this:
`uni_data = (uni_data-uni_train_mean)/uni_train_std`
This is done before:
```
x_train_uni, y_train_uni = univariate_data(uni_data, 0, TRAIN_SPLIT,
                                           univariate_past_history,
                                           univariate_future_target)
```
This is causing the past_history samples using values of future targets as well during the normalization. This is a bias. In reality, we cannot use future values to normalize current values.
This, I think, is a bias and a bug.

### Usage example

Normalization should be done after extraction of sequences and only using the LHS of the sequence. I still dont know if normalizing the RHS of the sequence is desired. but does not hurt as long as we denormalize

**Comment by Saduf2019** [comment]
@SarnathK 
This does not seem like a tensor flow issue.

Please provide details about what platform you are using (operating system, architecture). Also include your TensorFlow version. Also, did you compile from source or install a binary?

Make sure you also include the exact command if possible to produce the output included in your test case. If you are unclear what to include see the issue template displayed in [the Github new issue template](https://github.com/tensorflow/tensorflow/issues/new).

We ask for this in the issue submission template, because it is really difficult to help without that information. Thanks!

**Comment by SarnathK** [comment]
This is an issue in the Documentation. When I raised this issue, I remember selecting "documentation issue".

**Comment by SarnathK** [comment]
One of you, Can you at least confirm that this is actually
an issue?

**Comment by MarkDaoust** [comment]
Hi, 

Thanks for reporting this. You do have a valid point: all the training data is normalized based on all the training data, making the training performance unrealistic.

In some ways this could be dismissed as a non-issue, since this is the point of validation and test splits, to get a better view of real-world performance.

But it could be a good way to introduce moving averages, both for smoothing, or [ARIMA](https://otexts.com/fpp2/MA.html) style models.

**Comment by SarnathK** [comment]
Hi Mark,

Thanks for sharing your thoughts. Agree that validation and test should bring out the gap. But then, imho, it does not help the cause of creating a good model.... 

We ran into some unrealistic performance in our own project esp. for a longer multi-step model. We had used 6 horizons.

Apart from this, the way the sliding windows are created for training set, also introduces a bias because future values of 1 row will enter the history of another row creating some sort of bias while calculating mean and standard deviation. So using mean and deviation from training-set is even more confusing.

Finally we settled on slicing history_size of training data alone starting from t=0 (dataset[: , 0:history_size]) and using its mean and standard-deviation for the training-set and test-set. This ensures that the mean and standard-deviation are completely bias-free.

Finally, My request would be to include a note on this in the documentation so that readers understand the pitfalls of the other approaches.

Thanks,
Best,
Sarnath
